File: Neural-Networks-Micrograd/micrograd/variable.py
# Rewriting Value with gradient calculation
import numpy as np
import math

class Value:
    def __init__(self, data, _children=(), _op='', label=''):
        self.data = data
        self._prev = _children
        self.grad = 0.0
        self._backward = lambda: None
        self._op = _op
        self.label = label
        
    def __repr__(self):
        return f"Value(data={self.data}, label={self.label}, grad={self.grad})"
        
    def __add__(self, other):
        if np.isscalar(other):
            other = Value(other)
        out = Value(self.data + other.data, (self, other), '+')
        
        def _backward():
            self.grad += 1.0 * out.grad  # Local gradient is 1 for addition, which is multiplied by the incoming gradient from the output
            other.grad += 1.0 * out.grad
        out._backward = _backward

        return out

    def __radd__(self, other):
        return self+other

    def __sub__(self, other):
        return self + (-1*other)

    def __rsub__(self, other): # other - self
        return other + (-self)
    
    def __mul__(self, other):
        if np.isscalar(other):
            other = Value(other)
        out = Value(self.data * other.data, (self, other), '*')
        
        def _backward():
            self.grad += other.data*out.grad # Local derivative of multiplication is just the variable which is being multiplied to.
            other.grad += self.data*out.grad # Local derivative then multiplied with the incoming gradient from the output
        out._backward = _backward

        return out

    def __rmul__(self, other):
        return self*other

    def __truediv__(self, other):
        if np.isscalar(other):
            other = Value(other)
        return self*other**-1

    def __rtruediv__(self, other):
        return other*self**-1
    
    def __pow__(self, other):
        assert isinstance(other, (int, float))
        out = Value(self.data**other, (self,), f'**{other}')
        def _backward():
            self.grad = (other*self.data**(other-1))*out.grad   # Local derivative of power function chained with output grad
        out._backward = _backward
        return out

    def exp(self):
        n = self.data
        exp_value = math.exp(n)
        out = Value(exp_value, (self, ), 'exp')

        def __backward():
            self.grad = out.data*out.grad   # Local gradient of exponentiation is itself
        out._backward = __backward

        return out

    def tanh(self):
        n = self.data
        tanh_value = (math.exp(2*n)-1)/(math.exp(2*n)+1)
        out = Value(tanh_value, (self,), 'tanh')

        def _backward():
            self.grad += (1 - out.data**2)*out.grad
        out._backward = _backward

        return out

    def relu(self):
        n = self.data
        relu_value = n if n > 0 else 0
        out = Value(relu_value, (self, ), 'relu')

        def _backward():
            self.grad += (1.0*out.grad if n > 0 else 0) # Relu has zero local gradient when n <=0. Otherwise it is y=x function with local gradient 1.0
        out._backward = _backward

        return out

    def backward(self):
        topo = []
        def build_topo(v, visited):
            if v not in visited:
                visited.add(v)
                for child in v._prev:
                    build_topo(child, visited)
                topo.append(v)
        build_topo(self, set())
        self.grad = 1.0
        for node in reversed(topo):
            node._backward()

File: Neural-Networks-Micrograd/micrograd_cpp/include/micrograd_cpp/variable.cpp
#include <ostream>
#include <algorithm>
#include <cmath>
#include <string>
#include <vector>
#include <memory>
#include <functional>

using namespace std;

namespace micrograd_cpp
{
    class Variable;
    using VariablePtr = shared_ptr<Variable>;

    class Variable : public enable_shared_from_this<Variable>
    {
    public:
        float data;
        float grad;
        vector<VariablePtr> _prev;
        string op;
        string label;
        function<void()> backward;

        const vector<VariablePtr> &children() const { return _prev; }

        Variable(float data, vector<VariablePtr> _children = vector<VariablePtr>(), string op = "", string label = "")
        {
            this->data = data;
            this->grad = 0.0;
            this->_prev = _children;
            this->op = op;
            this->label = label;
            this->backward = []() {};
        }

        // function to print label when printing the object using cout
        friend ostream &operator<<(ostream &os, const Variable &v)
        {
            os << "Variable(data=" << v.data << ", grad=" << v.grad << ", label=" << v.label << ")";
            return os;
        }

        // Overload the + operator
        VariablePtr operator+(const VariablePtr other)
        {
            VariablePtr self = shared_from_this();
            vector<VariablePtr> children = vector<VariablePtr>{self, other};

            VariablePtr out = make_shared<Variable>(self->data + other->data, children, "+");
            out->backward = [out, other, self]()
            {
                self->grad += 1.0 * out->grad;
                other->grad += 1.0 * out->grad;
            };

            return out;
        }

        // Overload the * operator
        VariablePtr operator*(const VariablePtr &other)
        {
            VariablePtr out = make_shared<Variable>(this->data * other->data, vector<VariablePtr>{shared_from_this(), other}, "*");
            out->backward = [out, other, self = shared_from_this()]()
            {
                self->grad += other->data * out->grad;
                other->grad += self->data * out->grad;
            };

            return out;
        }

        // Define the pow operator
        VariablePtr pow(float exponent)
        {
            VariablePtr self = shared_from_this();
            VariablePtr out = make_shared<Variable>(std::pow(this->data, exponent), vector<VariablePtr>{self}, "^" + to_string(exponent));
            out->backward = [out, exponent, self]()
            {
                self->grad += exponent * std::pow(self->data, exponent - 1) * out->grad;
            };

            return out;
        }

        // Define the tanh function
        VariablePtr tanh()
        {
            float n = this->data;
            float tanh_value = (exp(2 * n) - 1) / (exp(2 * n) + 1);
            VariablePtr out = make_shared<Variable>(tanh_value, vector<VariablePtr>{shared_from_this()}, "tanh");
            out->backward = [out, self = shared_from_this()]()
            {
                self->grad += (1 - out->data * out->data) * out->grad;
            };

            return out;
        }

        // Define the relu function
        VariablePtr relu()
        {
            float n = this->data;
            float relu_value = n > 0 ? n : 0;
            VariablePtr out = make_shared<Variable>(relu_value, vector<VariablePtr>{shared_from_this()}, "relu");
            out->backward = [out, self = shared_from_this()]()
            {
                self->grad += (out->data > 0 ? 1 : 0) * out->grad;
            };

            return out;
        }
    };

    inline VariablePtr operator*(VariablePtr lhs, VariablePtr rhs)
    {
        VariablePtr out = make_shared<Variable>(lhs->data * rhs->data, vector<VariablePtr>{lhs, rhs}, "*");
        out->backward = [out, lhs, rhs]()
        {
            lhs->grad += rhs->data * out->grad;
            rhs->grad += lhs->data * out->grad;
        };

        return out;
    }

    inline VariablePtr operator*(VariablePtr lhs, float val)
    {
        return lhs * make_shared<Variable>(val);
    }

    inline VariablePtr operator*(float val, VariablePtr rhs)
    {
        return make_shared<Variable>(val) * rhs;
    }

    inline VariablePtr operator+(VariablePtr lhs, VariablePtr rhs)
    {
        VariablePtr out = make_shared<Variable>(lhs->data + rhs->data, vector<VariablePtr>{lhs, rhs}, "+");
        out->backward = [out, lhs, rhs]()
        {
            lhs->grad += out->grad;
            rhs->grad += out->grad;
        };
        return out;
    }

    inline VariablePtr operator+(VariablePtr lhs, float val)
    {
        return lhs + make_shared<Variable>(val);
    }

    inline VariablePtr operator-(VariablePtr rhs)
    {
        return rhs * make_shared<Variable>(-1.0);
    }

    inline VariablePtr operator-(VariablePtr lhs, VariablePtr rhs)
    {
        return lhs + (-rhs);
    }

    inline VariablePtr operator/(VariablePtr lhs, VariablePtr rhs)
    {
        return lhs * rhs->pow(-1);
    }

    inline VariablePtr operator/(VariablePtr lhs, float val)
    {
        return lhs / make_shared<Variable>(val);
    }

    inline VariablePtr operator/(float val, VariablePtr rhs)
    {
        return make_shared<Variable>(val) / rhs;
    }
}
